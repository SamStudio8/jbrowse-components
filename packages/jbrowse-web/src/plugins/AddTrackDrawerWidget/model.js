import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'

export default types.model('AddTrackModel', {
  id: ElementId,
  type: types.literal('AddTrackDrawerWidget'),
})
