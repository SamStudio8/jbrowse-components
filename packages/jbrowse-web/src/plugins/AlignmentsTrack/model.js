import {
  types,
  getSnapshot,
  getParent,
  getRoot,
  flow,
  isAlive,
  setLivelinessChecking,
} from 'mobx-state-tree'

import { autorun } from 'mobx'
import { ConfigurationReference, getConf } from '../../configuration'

import { Region } from '../../mst-types'
import { renderRegionWithWorker } from '../../render'

import { BaseTrack as LinearGenomeTrack } from '../LinearGenomeView/models/model'
import AlignmentsTrack, {
  AlignmentsTrackBlock,
} from './components/AlignmentsTrack'

setLivelinessChecking('error')

const BlockState = types
  .model('BlockState', {
    key: types.string,
    region: Region,
  })
  .volatile(() => ({
    filled: false,
    data: undefined,
    reactComponent: AlignmentsTrackBlock,
    html: '',
    error: undefined,
  }))
  .views(self => ({
    get rendererType() {
      const { pluginManager } = getRoot(self)
      const track = getParent(self, 2)
      const RendererType = pluginManager.getRendererType(track.rendererType)
      if (!RendererType)
        throw new Error(`renderer "${track.rendererType} not found`)
      if (!RendererType.ReactComponent)
        throw new Error(
          `renderer ${
            track.rendererType
          } has no ReactComponent, it may not be completely implemented yet`,
        )
      return RendererType
    },

    get renderProps() {
      const view = getParent(self, 4) // view -> [tracks] -> [blocks]
      return { bpPerPx: view.bpPerPx }
    },
  }))
  .actions(self => ({
    afterAttach() {
      self.render()
    },
    render: flow(function* renderBlock() {
      if (!isAlive(self)) return
      const track = getParent(self, 2)
      const view = getParent(track, 2)
      const root = getParent(view, 2)
      try {
        // console.log('calling', self.region.toJSON())
        const { html, ...data } = yield renderRegionWithWorker(root.app, {
          region: self.region,
          adapterType: track.adapterType.name,
          adapterConfig: getConf(track, 'adapter'),
          rendererType: track.rendererType,
          renderProps: self.renderProps,
          sessionId: track.id,
          timeout: 10000,
        })
        if (!isAlive(self)) return
        self.filled = true
        self.data = data
        self.html = html
        // console.log('finished', self.region.toJSON(), self.html)
      } catch (error) {
        // the rendering failed for some reason
        console.error(error)
        self.error = error
      }
    }),
  }))

const BlockBasedTrackState = types.compose(
  'BlockBasedTrackState',
  LinearGenomeTrack,
  types
    .model({
      blockState: types.map(BlockState),
    })
    .actions(self => {
      let blockWatchDisposer
      function disposeBlockWatch() {
        if (blockWatchDisposer) blockWatchDisposer()
        blockWatchDisposer = undefined
      }
      return {
        afterAttach() {
          const view = getParent(self, 2)
          // watch the parent's blocks to update our block state when they change
          blockWatchDisposer = autorun(() => {
            // create any blocks that we need to create
            const blocksPresent = {}
            view.blocks.forEach(block => {
              blocksPresent[block.key] = true
              if (!self.blockState.has(block.key))
                self.blockState.set(
                  block.key,
                  BlockState.create({
                    key: block.key,
                    region: block,
                  }),
                )
            })
            // delete any blocks we need to delete
            self.blockState.forEach((value, key) => {
              if (!blocksPresent[key]) self.blockState.delete(key)
            })
          })
        },
        beforeDetach: disposeBlockWatch,
        beforeDestroy: disposeBlockWatch,
      }
    }),
)

export default (pluginManager, configSchema) =>
  types.compose(
    'AlignmentsTrack',
    BlockBasedTrackState,
    types
      .model({
        type: types.literal('AlignmentsTrack'),
        configuration: ConfigurationReference(configSchema),
        // the renderer that the user has selected in the UI, empty string
        // if they have not made any selection
        selectedView: types.optional(types.string, ''),
        height: types.optional(types.integer, 100),
      })
      .volatile(self => ({
        reactComponent: AlignmentsTrack,
      }))
      // .actions(self => ({
      //   afterAttach() {
      //     onPatch(self, patch => {
      //       console.log('patch', self.name, patch)
      //     })
      //   },
      // }))
      .views(self => ({
        // the renderer type is based on the "view" selected in the UI: pileup, coverage, etc
        get rendererType() {
          const defaultView = getConf(self, 'defaultView')
          const viewName = self.selectedView || defaultView
          const rendererType = { pileup: 'PileupRenderer' }[viewName]
          if (!rendererType)
            throw new Error(`unknown alignments view name ${viewName}`)
          return rendererType
        },

        get adapterType() {
          const adapterConfig = getConf(self, 'adapter')
          if (!adapterConfig)
            throw new Error(
              `no adapter configuration provided for ${self.type}`,
            )
          const adapterType = pluginManager.getAdapterType(adapterConfig.type)
          if (!adapterType)
            throw new Error(`unknown adapter type ${adapterConfig.type}`)
          return adapterType
        },

        get adapter() {
          // TODO: refactor this someplace else to make this code shareable
          const adapter = new self.adapterType.AdapterClass(
            getSnapshot(self.configuration.adapter),
          )
          return adapter
        },
      })),
  )
