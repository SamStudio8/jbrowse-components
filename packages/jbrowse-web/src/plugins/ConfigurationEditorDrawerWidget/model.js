import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'
import { Assembly } from '../../rootModel'

export default pluginManager =>
  types
    .model('ConfigurationEditorDrawerWidget', {
      id: ElementId,
      type: types.literal('ConfigurationEditorDrawerWidget'),
      target: types.reference(
        types.union(
          pluginManager.pluggableConfigSchemaType('track'),
          pluginManager.pluggableConfigSchemaType('view'),
          Assembly,
        ),
      ),
    })
    // .volatile(() => ({
    //   target: undefined,
    // }))
    .actions(self => ({
      setTarget(newTarget) {
        self.target = newTarget
      },
    }))
//    .views(self => ({}))
