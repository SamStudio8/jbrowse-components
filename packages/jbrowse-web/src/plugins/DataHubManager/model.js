import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'

export default types.model('DataHubModel', {
  id: ElementId,
  type: types.literal('DataHubDrawerWidget'),
})
