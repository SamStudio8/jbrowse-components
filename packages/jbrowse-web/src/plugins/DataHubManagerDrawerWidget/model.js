import { types } from 'mobx-state-tree'
import { ElementId } from '@gmod/jbrowse-core/mst-types'

export default types.model('DataHubModel', {
  id: ElementId,
  type: types.literal('DataHubDrawerWidget'),
})
