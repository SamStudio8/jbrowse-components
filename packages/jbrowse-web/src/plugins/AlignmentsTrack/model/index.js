import { types } from 'mobx-state-tree'
import { ElementId } from '../../../mst-types'
import AlignmentsTrack from './alignmentsTrack'

export default AlignmentsTrack

export const AlignmentsFeatureDrawerWidgetModel = types
  .model('AlignmentsFeatureDrawerWidget', {
    id: ElementId,
    type: types.literal('AlignmentsFeatureDrawerWidget'),
  })
  .volatile(() => ({
    featureData: {},
  }))
  .actions(self => ({
    setFeatureData(data) {
      self.featureData = data
    },
    clearFeatureData() {
      self.featureData = {}
    },
  }))
