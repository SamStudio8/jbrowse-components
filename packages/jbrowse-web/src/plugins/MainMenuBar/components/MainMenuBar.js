import AppBar from '@material-ui/core/AppBar'
import { withStyles } from '@material-ui/core/styles'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import { values } from 'mobx'
import { observer, PropTypes as MobxPropTypes } from 'mobx-react'
import PropTypes from 'prop-types'
import React from 'react'
import DropDownMenu from './DropDownMenu'

const styles = {
  root: {
    flexGrow: 1,
  },
  grow: {
    flexGrow: 1,
  },
}

function MainMenuBar(props) {
  const { classes, model } = props

  return (
    <AppBar className={classes.root} position="static">
      <Toolbar variant="dense">
        {values(model.menus).map(menu => (
          <DropDownMenu
            key={menu.name}
            menuTitle={menu.name}
            menuItems={menu.menuItems}
            model={model}
          />
        ))}
        <div className={classes.grow} />
        <Typography variant="h6" color="inherit">
          JBrowse
        </Typography>
      </Toolbar>
    </AppBar>
  )
}

MainMenuBar.propTypes = {
  classes: PropTypes.shape({
    grow: PropTypes.shape.isRequired,
    root: PropTypes.shape.isRequired,
  }).isRequired,
  model: MobxPropTypes.observableObject.isRequired,
}

export default withStyles(styles)(observer(MainMenuBar))
