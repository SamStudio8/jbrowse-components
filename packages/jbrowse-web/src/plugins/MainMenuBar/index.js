import { lazy } from 'react'
import { ConfigurationSchema } from '../../configuration'
import Plugin, { DrawerWidgetType, MenuBarType } from '../../Plugin'
import {
  AboutDrawerWidgetModel,
  HelpDrawerWidgetModel,
  MainMenuBarModel,
} from './model'

const MainMenuBarComponent = lazy(() => import('./components/MainMenuBar'))
const AboutDrawerWidgetComponent = lazy(() =>
  import('./components/AboutDrawerWidget'),
)
const HelpDrawerWidgetComponent = lazy(() =>
  import('./components/HelpDrawerWidget'),
)

export default class MainMenuBar extends Plugin {
  install(pluginManager) {
    pluginManager.addMenuBarType(() => {
      const stateModel = MainMenuBarModel

      const configSchema = ConfigurationSchema('MainMenuBar', {})

      return new MenuBarType({
        name: 'MainMenuBar',
        configSchema,
        stateModel,
        LazyReactComponent: MainMenuBarComponent,
      })
    })

    pluginManager.addDrawerWidgetType(() => {
      const stateModel = AboutDrawerWidgetModel

      const configSchema = ConfigurationSchema('AboutDrawerWidget', {})

      return new DrawerWidgetType({
        name: 'AboutDrawerWidget',
        configSchema,
        stateModel,
        LazyReactComponent: AboutDrawerWidgetComponent,
      })
    })

    pluginManager.addDrawerWidgetType(() => {
      const stateModel = HelpDrawerWidgetModel

      const configSchema = ConfigurationSchema('HelpDrawerWidget', {})

      return new DrawerWidgetType({
        name: 'HelpDrawerWidget',
        configSchema,
        stateModel,
        LazyReactComponent: HelpDrawerWidgetComponent,
      })
    })
  }
}
