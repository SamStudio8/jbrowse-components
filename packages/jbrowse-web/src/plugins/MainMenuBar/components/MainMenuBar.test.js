import { createShallow } from '@material-ui/core/test-utils'
import React from 'react'
import { MainMenuBarModel } from '../model'
import MainMenuBar from './MainMenuBar'

describe('<MainMenuBar />', () => {
  let shallow

  beforeAll(() => {
    shallow = createShallow()
  })

  it('renders', () => {
    const menubar = MainMenuBarModel.create({
      id: 'testingId',
      type: 'MainMenuBar',
    })
    const wrapper = shallow(
      <MainMenuBar
        model={menubar}
        rootModel={{ activeDrawerWidgets: new Map() }}
      />,
    )
      .first()
      .shallow()
    expect(wrapper).toMatchSnapshot()
  })
})
