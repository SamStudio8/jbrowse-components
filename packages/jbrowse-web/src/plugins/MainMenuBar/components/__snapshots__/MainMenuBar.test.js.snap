// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainMenuBar /> renders 1`] = `
<WithStyles(AppBar)
  className="MainMenuBar-root-1"
  position="static"
>
  <WithStyles(Toolbar)
    variant="dense"
  >
    <WithStyles(Typography)
      color="inherit"
      variant="h6"
    >
      JBrowse
    </WithStyles(Typography)>
    <div
      className="MainMenuBar-grow-2"
    />
    <Component
      key="Help"
      menuItems={
        Array [
          Object {
            "callback": "openAbout",
            "icon": "info",
            "name": "About",
          },
          Object {
            "callback": "openHelp",
            "icon": "help",
            "name": "Help",
          },
          Object {
            "callback": "openAssemblyEditor",
            "icon": "edit",
            "name": "Assemblies",
          },
        ]
      }
      menuTitle="Help"
    />
    <div
      style={
        Object {
          "width": 0,
        }
      }
    />
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;
